catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/security/ShareableMetacardPolicyExtension.java

```diff
@@ -125,14 +134,26 @@ private KeyValueCollectionPermission isPermitted(
 
     // filter out all implied permissions
     Function<Set<String>, KeyValueCollectionPermission> filterPermissions =
-        (implied) -> {
-          List<KeyValuePermission> values =
-              permissions
+        (impliedKeys) -> {
+          KeyValueCollectionPermission implied =
+              new KeyValueCollectionPermission(
+                  match.getAction(),
+                  impliedKeys
+                      .stream()
+                      .map(key -> new KeyValuePermission(key, grouped.get(key)))
+                      .collect(Collectors.toList()));
+
+          List<Permission> remaining =
+              allPerms
```

Comment:
I don't think you want to use the allPerms here. That is the unsorted list of matchone and matchall permissions. If you try to imply things out of this list and return them, you are going to mix up the remaining permissions that the PDP has left to imply. That unsorted list is only passed in as a reference so that you could potentially key off of the existence of one or more permissions if that needed to be part of your decision. You would also not be able to handle matchOne and matchAll permissions with the same method. If you take a look at the PDP class: https://github.com/codice/ddf/blob/master/platform/security/pdp/security-pdp-authzrealm/src/main/java/ddf/security/pdp/realm/AuthzRealm.java#L267 you'll see that it uses a special collection for matchOne that can flip the implied values around to match just a single value instead of all of them. You'd need to do something similar here if you're trying to use the same method and check if the perms imply each other below in the stream.

Reply:
https://github.com/codice/ddf/blob/master/platform/security/platform-security-core-api/src/main/java/ddf/security/permission/MatchOneCollectionPermission.java#L19-L23